onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlSTFT.h

```diff
@@ -0,0 +1,610 @@
+        {
+            DML_BUFFER_BINDING bufferBinding = { m_framingOperator.persistentResource.Get(), 0, persistentBufferSize };
+            DML_BINDING_DESC bindingDesc = { DML_BINDING_TYPE_BUFFER, &bufferBinding };
+            m_framingOperator.bindingTable->BindPersistentResource(&bindingDesc);
+        }
+
+        // Transition resources COMMON -> UAV
+        commandList->ResourceBarrier(barrierCount, barriers.data());
+
+        m_framingOperator.commandRecorder->RecordDispatch(
+            commandList,
+            m_framingOperator.op.Get(),
+            m_framingOperator.bindingTable.Get()
+        );
+
+        // Transition resources UAV -> COMMON
+        for (uint32_t barrierIndex = 0; barrierIndex < barrierCount; barrierIndex++)
+        {
+            std::swap(barriers[barrierIndex].Transition.StateBefore, barriers[barrierIndex].Transition.StateAfter);
+        }
+
+        commandList->ResourceBarrier(barrierCount, barriers.data());
+    }
+};
+
+struct STFTShapeInferrer : public WRL::Base<IMLOperatorShapeInferrer>
+{
+    STDMETHOD(InferOutputShapes)(IMLOperatorShapeInferenceContext* context) noexcept
+    {
+        try
+        {
+            ComPtr<IMLOperatorShapeInferenceContextPrivate> contextPrivate;
+            ORT_THROW_IF_FAILED(context->QueryInterface(IID_PPV_ARGS(&contextPrivate)));
+
+            DmlSTFTShapeInferenceParams info(contextPrivate.Get(), contextPrivate.Get());
+
+            std::array<uint32_t, 4> outputDims = { info.batchSize, info.frameCount, info.frameDftElementCount, 2 };
+
+            ORT_THROW_IF_FAILED(context->SetOutputTensorShape(0, onnxruntime::narrow<uint32_t>(outputDims.size()), outputDims.data()));
+        }
+        catch (...)
+        {
+            return E_FAIL;
+        }
+
+        return S_OK;
+    }
+};
+
+class DmlSTFTOperatorFactory : public WRL::Base<IMLOperatorKernelFactory>
+{
+public:
+    STDMETHOD(CreateKernel)(
+        IMLOperatorKernelCreationContext* context,
+        IMLOperatorKernel** kernel)
+    {
+        try
+        {
+            auto dftOperator = wil::MakeOrThrow<DmlSTFTOperator>(context);
+            dftOperator.CopyTo(kernel);
+            return S_OK;
+        }
+        catch (...)
+        {
+            return E_FAIL;
+        }
+    }
+
+    static void RegisterSTFTKernel(IMLOperatorRegistry* registry)
+    {
+        MLOperatorKernelDescription kernelDescription = {};
+        kernelDescription.domain = "";
+        kernelDescription.name = "STFT";
+        kernelDescription.minimumOperatorSetVersion = 17;
+        kernelDescription.executionType = MLOperatorExecutionType::D3D12;
+
+        // T1: tensor(float16), tensor(float), tensor(double), tensor(bfloat16)
+        MLOperatorEdgeTypeConstrant t1Constraint;
+        t1Constraint.typeLabel = "T1";
+        std::array<MLOperatorEdgeDescription, 2> t1AllowedEdges
+        {
+            MLOperatorEdgeDescription { MLOperatorEdgeType::Tensor, (uint64_t)MLOperatorTensorDataType::Float16 },
```

Comment:
>(uint64_t) [](http://example.com/codeflow?start=68&length=10)

[nit] Propose `uint64_t(MLOperatorTensorDataType::Float16)` just because C-style casts can be dangerous in the degree of castiness they perform. (or static_cast like line 569)

Reply:
Sorry. Copy-pasta from existing DmlDFT.h